src/AppInstallerRepositoryCore/Microsoft/CheckpointIndex.cpp

```diff
@@ -0,0 +1,212 @@
+
+        try
+        {
+            if (std::filesystem::exists(indexPath))
+            {
+                if (std::filesystem::is_regular_file(indexPath))
+                {
+                    try
+                    {
+                        AICLI_LOG(Repo, Info, << "Opening existing checkpoint index");
+                        return std::make_shared<CheckpointIndex>(CheckpointIndex::Open(indexPath.u8string(), openDisposition));
+                    }
+                    CATCH_LOG();
+                }
+
+                AICLI_LOG(Repo, Info, << "Attempting to delete bad checkpoint index file");
+                std::filesystem::remove_all(indexPath);
+            }
+
+            return std::make_shared<CheckpointIndex>(CheckpointIndex::CreateNew(indexPath.u8string()));
+        }
+        CATCH_LOG();
+
+        return {};
+    }
+
+#ifndef AICLI_DISABLE_TEST_HOOKS
+    std::optional<std::filesystem::path> s_CheckpointIndexDirectoryOverride{};
+    void TestHook_SetCheckpointIndexDirectory_Override(std::optional<std::filesystem::path>&& checkpointIndexDirectory)
+    {
+        s_CheckpointIndexDirectoryOverride = std::move(checkpointIndexDirectory);
+    }
+#endif
+
+    std::filesystem::path CheckpointIndex::GetCheckpointIndexPath(GUID guid)
+    {
+        wchar_t checkpointGuid[256];
+        THROW_HR_IF(E_UNEXPECTED, !StringFromGUID2(guid, checkpointGuid, ARRAYSIZE(checkpointGuid)));
+
+        const auto DefaultIndexDirectoryPath = Runtime::GetPathTo(Runtime::PathName::LocalState) / s_Checkpoints;
+#ifndef AICLI_DISABLE_TEST_HOOKS
+        const auto checkpointIndexDirectory = s_CheckpointIndexDirectoryOverride.has_value() ? s_CheckpointIndexDirectoryOverride.value() : DefaultIndexDirectoryPath;
+#else
+        const auto checkpointIndexDirectory = DefaultIndexDirectoryPath;
+#endif
+
+        if (!std::filesystem::exists(checkpointIndexDirectory))
+        {
+            std::filesystem::create_directories(checkpointIndexDirectory);
+            AICLI_LOG(Repo, Info, << "Creating checkpoint index directory: " << checkpointIndexDirectory);
+        }
+        else
+        {
+            THROW_HR_IF(ERROR_CANNOT_MAKE, !std::filesystem::is_directory(checkpointIndexDirectory));
+        }
+
+        auto indexPath = checkpointIndexDirectory / checkpointGuid;
+        indexPath.replace_extension(".db");
+        return indexPath;
+    }
+
+    bool CheckpointIndex::IsEmpty()
+    {
+        return m_interface->IsEmpty(m_dbconn);
+    }
+
+    CheckpointIndex::IdType CheckpointIndex::SetClientVersion(std::string_view clientVersion)
+    {
+        std::lock_guard<std::mutex> lockInterface{ *m_interfaceLock };
+        AICLI_LOG(Repo, Verbose, << "Setting client version [" << clientVersion << "]");
+
+        SQLite::Savepoint savepoint = SQLite::Savepoint::Create(m_dbconn, "checkpointindex_setclientversion");
+
+        IdType result = m_interface->SetClientVersion(m_dbconn, clientVersion);
```

Comment:
Similar to the other comment, it is much less work to have a generic method rather than piping in a new one every time you need a new value saved.